packages/react-async/src/component.tsx

```diff
@@ -1,135 +1,330 @@
+      },
+      [load, customKeepFresh],
     );
   }
 
-  function KeepFresh(props: KeepFreshProps & ConstantProps) {
-    const [componentProps, asyncProps] = splitProps(props);
+  function Preload(options: PreloadOptions) {
+    useIdleCallback(usePreload(options));
+    return null;
+  }
 
-    return (
-      <>
-        {renderKeepFresh(componentProps)}
-        <Async defer={DeferTiming.Idle} load={load} {...asyncProps} />
-      </>
+  Preload.displayName = `Async.Preload(${displayName})`;
+
+  function Prefetch(options: PrefetchOptions) {
+    const prefetch = usePrefetch(options);
+
+    useEffect(
+      () => {
+        prefetch();
+      },
+      [prefetch],
     );
+
+    return null;
+  }
+
+  Prefetch.displayName = `Async.Prefetch(${displayName})`;
+
+  function KeepFresh(options: KeepFreshOptions) {
+    useIdleCallback(useKeepFresh(options));
+    return null;
   }
 
+  KeepFresh.displayName = `Async.KeepFresh(${displayName})`;
+
   // Once we upgrade past TS 3.1, this will no longer be necessary,
   // because you can statically assign values to functions and TS
   // will know to augment its type
   const FinalComponent: AsyncComponentType<
+    ComponentType<Props>,
     Props,
-    PreloadProps,
-    PrefetchProps,
-    KeepFreshProps
-  > = AsyncComponent as any;
+    PreloadOptions,
+    PrefetchOptions,
+    KeepFreshOptions
+  > = Async as any;
+
+  Reflect.defineProperty(FinalComponent, 'resolver', {
+    value: resolver,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'Preload', {
+    value: Preload,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'Prefetch', {
+    value: Prefetch,
+    writable: false,
+  });
 
-  FinalComponent.Preload = Preload;
-  FinalComponent.Prefetch = Prefetch;
-  FinalComponent.KeepFresh = KeepFresh;
+  Reflect.defineProperty(FinalComponent, 'KeepFresh', {
+    value: KeepFresh,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'usePreload', {
+    value: usePreload,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'usePrefetch', {
+    value: usePrefetch,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'useKeepFresh', {
+    value: useKeepFresh,
+    writable: false,
+  });
 
   return FinalComponent;
 }
 
+function noopUse() {
+  return () => {};
+}
+
 function noopRender() {
   return null;
 }
 
-function splitProps<Props>(
-  props: Props & ConstantProps,
-): [Props, AsyncPropsRuntime] {
-  const {async, ...rest} = props as any;
-  return [rest, async];
+const DEFAULT_DISPLAY_NAME = 'Component';
+const FILENAME_REGEX = /([^/]*)\.\w+$/;
+
+function displayNameFromId(id?: string) {
+  if (!id) {
+    return DEFAULT_DISPLAY_NAME;
+  }
+
+  const match = id.match(FILENAME_REGEX);
+  return match ? match[1] : DEFAULT_DISPLAY_NAME;
+}
+
+function defaultRenderError(error: Error) {
+  if (error) {
+    throw error;
+  }
+
+  return null;
+}
+
+function Loader<T>({
```

Comment:
This thing might be worth testing in isolation / being in it's own file /shrug

Reply:
No, it’s an implementation detail. All of this is tested in the tests for the actual public API